include/verilated_trace_imp.cpp

```diff
@@ -90,6 +90,221 @@ class VerilatedTraceCallInfo {
     ~VerilatedTraceCallInfo() {}
 };
 
+#ifdef VL_TRACE_THREADED
+//=========================================================================
+// Buffer management
+
+template <> VerilatedTraceEntry* VerilatedTrace<VL_DERIVED_T>::getTraceBuffer() {
+    VerilatedTraceEntry* bufferp;
+    // Some jitter is expected, so some number of alternative trace buffers are
+    // required, but don't allocate more than 8 buffers.
+    if (m_numTraceBuffers < 8) {
+        // Allocate a new buffer if none is available
+        if (!m_buffersFromWorker.tryGet(bufferp)) {
+            m_numTraceBuffers++;
+
+            // We need to be able to store a new value for each signal, which is
+            // 'nextCode()' entries, plus up to 3 more words of metadata per signal,
+            // Plus fixed overhead of 1 for a termination flag and 2 for a time
+            // stamp update.
+            bufferp = new VerilatedTraceEntry[nextCode() + numSignals() * 3 + 3];
+        }
+    } else {
+        // Block until a buffer becomes available
+        bufferp = m_buffersFromWorker.get();
+    }
+    return bufferp;
+}
+
+template <> void VerilatedTrace<VL_DERIVED_T>::waitForBuffer(const VerilatedTraceEntry* buffp) {
+    // Collect buffers from worker and stash them until we get the one we want
+    std::deque<VerilatedTraceEntry*> stash;
+    do { stash.push_back(m_buffersFromWorker.get()); } while (stash.back() != buffp);
+    // Now put them back in the queue, in the original order
+    while (!stash.empty()) {
+        m_buffersFromWorker.put_front(stash.back());
+        stash.pop_back();
+    }
+}
+
+//=========================================================================
+// Worker thread
+
+#if 0
+#define VL_TRACE_THREAD_DEBUG(msg) std::cout << "TRACE THREAD: " << msg << std::endl
+#else
+#define VL_TRACE_THREAD_DEBUG(msg)
+#endif
```

Comment:
Suggest please move to closer to top of file (near defines) so more obvious this can be set.